executor/window_parallel.go

```diff
@@ -0,0 +1,367 @@
+	for i := range e.workers {
+		for range e.workers[i].inputCh {
+		}
+	}
+	for range e.outputCh {
+	}
+	e.executed = false
+
+	if e.runtimeStats != nil {
+		e.runtimeStats.SetConcurrencyInfo("WindowParallelConcurrency", e.concurrency)
+	}
+
+	return errors.Trace(e.baseExecutor.Close())
+}
+
+func recoveryWindowParallelExec(output chan *windowParallelOutput, r interface{}) {
+	err := errors.Errorf("%v", r)
+	output <- &windowParallelOutput{err: errors.Errorf("%v", r)}
+	logutil.BgLogger().Error("window parallel panicked", zap.Error(err))
+}
+
+func (e *WindowParallelExec) execDataFetcherThread(ctx context.Context) {
+	var (
+		ok        bool
+		err       error
+		groupKeys [][]byte
+	)
+	results := make([]*chunk.Chunk, len(e.workers))
+
+	defer func() {
+		if r := recover(); r != nil {
+			recoveryWindowParallelExec(e.outputCh, r)
+		}
+		for i := range e.workers {
+			close(e.workers[i].inputCh)
+		}
+	}()
+	for {
+		chk := newFirstChunk(e.children[0])
+		err = Next(ctx, e.children[0], chk)
+		if err != nil {
+			e.outputCh <- &windowParallelOutput{err: err}
+			return
+		}
+		if chk.NumRows() == 0 {
+			break
+		}
+
+		groupKeys, err = getGroupKey(e.ctx, chk, groupKeys, e.groupByItems)
+		if err != nil {
+			e.outputCh <- &windowParallelOutput{err: err}
+			return
+		}
+		numRows := chk.NumRows()
+		for i := 0; i < numRows; i++ {
+			workerIdx := int(murmur3.Sum32(groupKeys[i])) % len(e.workers)
+			w := &e.workers[workerIdx]
+
+			if results[workerIdx] == nil {
+				select {
+				case <-e.finishCh:
+					return
+				case results[workerIdx], ok = <-w.inputHolderCh:
+					if !ok {
+						return
+					}
+				}
+			}
+			results[workerIdx].AppendRow(chk.GetRow(i))
+			if results[workerIdx].IsFull() {
+				w.inputCh <- results[workerIdx]
+				results[workerIdx] = nil
+			}
+		}
+	}
+	for i := range e.workers {
+		if results[i] != nil {
+			e.workers[i].inputCh <- results[i]
+			results[i] = nil
+		}
+	}
+}
+
+func (e *WindowParallelExec) prepare4ParallelExec(ctx context.Context) {
+	go e.execDataFetcherThread(ctx)
+
+	waitGroup := &sync.WaitGroup{}
+	waitGroup.Add(len(e.workers))
+	for i := range e.workers {
+		go e.workers[i].run(ctx, waitGroup)
+	}
+
+	go e.waitWorkerAndCloseOutput(waitGroup)
+}
+
+func (e *WindowParallelExec) waitWorkerAndCloseOutput(waitGroup *sync.WaitGroup) {
+	waitGroup.Wait()
+	close(e.outputCh)
+}
+
+// Next implements the Executor Next interface.
+func (e *WindowParallelExec) Next(ctx context.Context, req *chunk.Chunk) error {
+	req.Reset()
+	if !e.prepared {
+		e.prepare4ParallelExec(ctx)
+		e.prepared = true
+	}
+
+	failpoint.Inject("windowParallelError", func(val failpoint.Value) {
+		if val.(bool) {
+			failpoint.Return(errors.New("WindowParallelExec.Next error"))
+		}
+	})
+
+	if e.executed {
+		return nil
+	}
+	for {
+		result, ok := <-e.outputCh
+		if !ok {
+			e.executed = true
+			return nil
+		}
+		if result.err != nil {
+			return result.err
+		}
+		req.SwapColumns(result.chk)
+		result.chk.Reset()
```

Comment:
Is it a redundant call? Since the chunk will be reset in `Next(ctx, w.windowExec, chk)`.